blog/2021-11-04-introducing-pwa.md

```diff
@@ -0,0 +1,168 @@
+
+The `@dhis2/app-runtime` package provides a React API to use a cacheable section that’s comprised of a `useCacheableSection(id)` hook and a `<CacheableSection id="">` component. In an _extremely_ simplified hypothetical version of the Dashboard app, the usage looks like this:
+
+```jsx
+import { useCacheableSection, CacheableSection } from "@dhis2/app-runtime";
+import { Button } from "@dhis2/ui";
+import { Dashboard } from "./dashboard.js";
+import { LoadingMask } from "./loading-mask.js";
+
+export function CacheableDashboard({ id }) {
+  const { startRecording } = useCacheableSection(id);
+
+  return (
+    <>
+      <Button onClick={() => startRecording()}>Make available offline</Button>
+      <CacheableSection id={id} loadingMask={<LoadingMask />}>
+        <Dashboard id={id} />
+      </CacheableSection>
+    </>
+  );
+}
+```
+
+When the “Make available offline” button is clicked and `startRecording` is called, the children of the `<CacheableSection>` component with the same `id` as the `useCacheableSection` hook will reload, and the provided loading mask will be rendered while the cacheable section records the requests that the section makes. Once the recording is done, that section can be accessed while offline!
+
+To learn more about the usage of cacheable sections, including an example of a Loading Mask, check out the [“Usage” section](https://runtime.dhis2.nu/#/advanced/offline?id=usage) in the App Runtime docs.
+
+### Online status
+
+If you took a look at the 2.37 version of the Dashboard app while offline, you may have noticed that many features of the app are disabled while offline, and some items in the interface change appearance.
+
+The App Runtime provides a `useOnlineStatus` hook that can access the online or offline status of the app, which the Dashboard app uses to enable or disable those features. Using the hook looks like this:
+
+```jsx
+import { useOnlineStatus } from "@dhis2/app-runtime";
+
+const OnlineComponent = () => {
+  const { online, offline } = useOfflineStatus();
+
+  return <p>{online ? "Online" : "Offline"}</p>;
+};
+```
+
+## An important caveat
+
+These features are exciting and the opportunity to take apps offline can be very useful when taken into areas with low network connectivity, but there is an important limitation that you should know before using the PWA features.
+
+In its current implementation, **cached data is not encrypted** when stored offline, and a malicious actor could inspect access and inspect the cached data if they gain access to a user’s device. For this reason, it’s important to **not use** PWA features in an app that will handle sensitive data, especially if the app is likely to be used on shared devices.
```

Comment:
```suggestion
In its current implementation, **cached data is not encrypted** when stored offline, and a malicious actor could inspect access and inspect the cached data if they gain access to a user’s device. For this reason, it is **not recommended** to use PWA features in an app that will handle sensitive data, especially if the app is likely to be used on shared devices.  Some protections are currently provided by the platform to guard against unauthorized access to data when a user logs out or when a new user logs in, but without encryption it might still be possible for a savvy attacker to access some offline data.  Importantly, the data cached by the PWA is the same as might be cached at the browser layer in a non-PWA application, and so the risk of data exposure has not significantly increased.
```